// angular-phone-number-input/esm2022/lib/angular-phone-number-input.component.mjs
import { Component, Input, forwardRef, inject } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, NG_VALIDATORS, NG_VALUE_ACCESSOR, Validators, } from '@angular/forms';
import { DomSanitizer } from '@angular/platform-browser';
import { Countries } from '../assets/country-dataset';
import { globeSVG } from '../assets/globe-svg';
import { isValidPhoneNumber } from 'libphonenumber-js';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@angular/forms";
export class AngularPhoneNumberInput {
    constructor() {
        this.error = false;
        this.border = true;
        this.sanitizer = inject(DomSanitizer);
        this.countryList = Countries;
        this.dataSet = Countries;
        this.showModal = false;
        this.inputFocused = false;
        this.disabled = false;
        this.globeSVG = globeSVG;
        /** Callback function triggered on value change. */
        this.onChange = () => { };
        /** Callback function triggered on input touch. */
        this.onTouched = () => { };
        /** Handles the input blur event. */
        this.onBlur = () => {
            this.inputFocused = false;
        };
        /**
         * Handles the input focus event.
         */
        this.onFocus = () => {
            this.inputFocused = true;
        };
        /**
         * Writes a new value to the input.
         * @param obj - The new value.
         */
        this.writeValue = (value) => {
            if (!!value) {
                this.extractValues(value);
            }
        };
        /**
         * Registers a callback function to be executed when the input value changes.
         * @param fn - The callback function.
         */
        this.registerOnChange = (fn) => {
            this.onChange = fn;
        };
        /**
         * Registers a callback function to be executed when the input is touched.
         * @param fn - The callback function.
         */
        this.registerOnTouched = (fn) => {
            this.onTouched = fn;
        };
        /**
         * Sets the disabled state of the input.
         * @param isDisabled - The disabled state.
         */
        this.setDisabledState = (isDisabled) => {
            this.disabled = isDisabled;
        };
        /**
         * Sets the default country based on the provided defaultCountry value.
         * If the default country exists in the countryList, sets the selectedCountry accordingly.
         */
        this.setDefaultCountry = () => {
            const defCountryData = this.countryList[this.defaultCountry];
            this.selectedCountry = defCountryData ?? null;
        };
        /**
         * Sets preferred countries based on the provided preferredCountries array.
         * Filters the countryList to contain only the preferred countries if they exist.
         */
        this.setPreferredCountries = () => {
            if (this.preferredCountries && this.preferredCountries.length > 0) {
                // Filter the countries to include only preferred countries
                const preferredCountryList = Object.fromEntries(Object.entries(Countries).filter(([code]) => this.preferredCountries.includes(code)));
                // If preferred countries are found, update the countryList
                if (Object.keys(preferredCountryList).length > 0) {
                    this.countryList = preferredCountryList;
                }
            }
        };
        /**
         * Sets the full phone number by concatenating the country dial code and the input value.
         * Updates the value using the onChange callback.
         * @returns {void}
         */
        this.setFullPhoneNumber = () => {
            const fullPhoneNumber = this.selectedCountry ? this.selectedCountry.dialCode + (this.value || '') : this.value || '';
            // Trigger the onChange callback with the updated full phone number
            this.onChange(fullPhoneNumber);
        };
        /**
         * Extracts the country code and phone number from the provided phone number string.
         * @param {string} phoneNumber - The input phone number string.
         * @returns {void}
         */
        this.extractValues = (phoneNumber) => {
            let country; // Variable to store the extracted country data
            const countryCodes = Object.keys(Countries);
            // Sort the country codes by length in descending order
            const sortedCodes = countryCodes.sort((a, b) => Countries[b].dialCode.length - Countries[a].dialCode.length);
            // Iterate through sorted country codes to find a match in the input value
            for (const code of sortedCodes) {
                if (phoneNumber.startsWith(Countries[code].dialCode)) {
                    country = Countries[code];
                    this.value = phoneNumber.substring(country.dialCode.length);
                    this.selectedCountry = country;
                    return;
                }
            }
            // If no matching country code is found, attempt to extract country code based on length (1 to 4 digits)
            for (let i = 1; i <= 4; i++) {
                const potentialCode = phoneNumber.substring(0, i);
                const matchingCode = countryCodes.find(codeKey => Countries[codeKey].dialCode === `+${potentialCode}`);
                if (matchingCode) {
                    country = Countries[matchingCode];
                    this.value = phoneNumber.substring(i);
                    this.selectedCountry = country;
                    return;
                }
            }
            // If no country code is identified, consider the entire input as the phone number
            this.value = phoneNumber.replace('+', '');
            return;
        };
        /**
         * Retrieves the sanitized HTML representation of an icon string.
         * @param {string | undefined} iconString - The string representation of an SVG icon.
         * @returns {SafeHtml | undefined} The sanitized HTML representation of the icon string.
         */
        this.getIcon = (iconString) => {
            if (iconString) {
                return this.sanitizer.bypassSecurityTrustHtml(iconString);
            }
            return undefined;
        };
        /**
         * Validates the input control.
         * @param control - The AbstractControl instance.
         * @returns ValidationErrors if value is null.
         */
        this.validate = (control) => {
            const fullPhoneNumber = this.selectedCountry ? this.selectedCountry.dialCode + (this.value || '') : '+' + this.value || '';
            // if no value update and inputted control has validation with required or return true
            if (!this.value || this.value === '') {
                if (control.hasValidator(Validators.required))
                    return { required: true };
                return null;
            }
            // check valid phone number or not with libphonenumber-js
            if (!isValidPhoneNumber(fullPhoneNumber.toString())) {
                return { status: "INVALID" };
            }
            return null;
        };
    }
    /**
     * Initializes default country and preferred countries on component initialization.
     * Calls setDefaultCountry and setPreferredCountries methods.
     */
    ngOnInit() {
        this.setDefaultCountry();
        this.setPreferredCountries();
    }
    /**
     * Toggles the modal visibility and sets the dataSet to the current countryList.
     */
    toggleModal() {
        this.dataSet = this.countryList;
        this.showModal = !this.showModal;
    }
    /**
     * Selects a country and emits the corresponding event.
     * @param {Country} data - The selected country data.
     */
    selectCountry(data) {
        this.selectedCountry = data.value;
        this.setFullPhoneNumber();
        this.toggleModal();
    }
    /**
     * Tries to determine the selected country based on the provided input value.
     * Updates the selected country and value if a matching country code is found.
     * @returns {void}
     */
    getCountryWithInput() {
        if (this.value) {
            // temp number to hold the input value
            const tempNum = '+' + this.value.toString();
            if (!this.selectedCountry && tempNum.length === 5) {
                const countryCodes = Object.keys(Countries);
                const sortedCodes = countryCodes.sort((a, b) => Countries[b].dialCode.length - Countries[a].dialCode.length);
                for (const code of sortedCodes) {
                    if (tempNum.startsWith(Countries[code].dialCode)) {
                        this.selectedCountry = Countries[code];
                        // remove the country code from input value
                        this.value = this.value.toString().substring(this.selectedCountry.dialCode.length - 1);
                        break;
                    }
                }
            }
        }
    }
    /**
     * Emits the current input value.
     */
    onInputChanged() {
        this.setFullPhoneNumber();
        this.getCountryWithInput();
    }
    /**
     * Filters the countries based on the provided search term and updates the dataset.
     * @param {string} searchTerm - The term to search for within country names or dial codes.
     * @returns {void}
     */
    filterCountries(searchTerm) {
        // Convert the searchTerm to lowercase for case-insensitive matching
        searchTerm = searchTerm.toLowerCase();
        // Filter the countryList based on the searchTerm and update dataSet
        this.dataSet = Object.keys(this.countryList)
            .filter(code => this.countryList[code].country.toLowerCase().includes(searchTerm) ||
            this.countryList[code].dialCode.includes(searchTerm))
            .reduce((obj, code) => {
            // Reduce the filtered keys back into an object
            obj[code] = this.countryList[code];
            return obj;
        }, {});
    }
    /**
     * Handles changes in the search input by invoking the filterCountries method with the input value.
     * @param {Event} event - The event object from the search input.
     * @returns {void}
     */
    onSearchChange(event) {
        const searchTerm = event.target.value;
        this.filterCountries(searchTerm);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.12", ngImport: i0, type: AngularPhoneNumberInput, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "17.3.12", type: AngularPhoneNumberInput, isStandalone: true, selector: "angular-phone-number-input", inputs: { error: "error", defaultCountry: "defaultCountry", preferredCountries: "preferredCountries", border: "border" }, providers: [
            {
                provide: NG_VALUE_ACCESSOR,
                useExisting: forwardRef(() => AngularPhoneNumberInput),
                multi: true,
            },
            {
                provide: NG_VALIDATORS,
                useExisting: forwardRef(() => AngularPhoneNumberInput),
                multi: true,
            },
        ], ngImport: i0, template: "<div class=\"wrap flex align-center\" [class.border]=\"border\" [class.active]=\"inputFocused\" [class.error]=\"error\">\n    <div class=\"flex\">\n        @if(selectedCountry){\n        <div class=\"flex align-center\" (click)=\"toggleModal()\">\n            <span class=\"flag\" [innerHTML]=\"getIcon(selectedCountry.flag)\"></span>\n\n            <svg class=\"ml-5\" xmlns=\"http://www.w3.org/2000/svg\" width=\"10\" height=\"6\" viewBox=\"0 0 10 6\" fill=\"none\">\n                <path d=\"M4.99959 5.13698L0.725586 0.862976H9.27359L4.99959 5.13698Z\" fill=\"black\" fill-opacity=\"0.6\" />\n            </svg>\n        </div>\n        <span class=\"text ml-5\">{{selectedCountry.dialCode}}</span>\n        }@else{\n        <div class=\"flex align-center\" (click)=\"toggleModal()\">\n            <span class=\"flag\" [innerHTML]=\"getIcon(globeSVG)\"></span>\n            <svg class=\"ml-5\" xmlns=\"http://www.w3.org/2000/svg\" width=\"10\" height=\"6\" viewBox=\"0 0 10 6\" fill=\"none\">\n                <path d=\"M4.99959 5.13698L0.725586 0.862976H9.27359L4.99959 5.13698Z\" fill=\"black\" fill-opacity=\"0.6\" />\n            </svg>\n        </div>\n        }\n\n    </div>\n\n    <input class=\"input-box text\" type=\"number\" placeholder=\"\" [(ngModel)]=\"value\" [disabled]=\"disabled\"\n        (input)=\"onChange(value); onInputChanged()\" (blur)=\"onBlur(); onTouched()\" (focus)=\"onFocus()\" />\n\n    @if(showModal){\n    <div class=\"modal-wrap\">\n        <div>\n            <input class=\"search-box\" (input)=\"onSearchChange($event)\" placeholder=\"Search\" />\n        </div>\n        <div class=\"country-list\">\n            <div style=\"display: flex;\" *ngFor=\"let data of dataSet | keyvalue\">\n                <div class=\"modal-country-wrap flex\" (click)=\"selectCountry(data)\">\n                    <span class=\"flag\" [innerHTML]=\"getIcon(data.value.flag)\"></span>\n                    <span class=\"modal-country ml-5\">{{data.value.dialCode}}</span>\n                    <span class=\"modal-country ml-5\">{{data.value.country}}</span>\n                </div>\n            </div>\n        </div>\n    </div>\n    }\n</div>", styles: ["input[type=number]::-webkit-inner-spin-button,input[type=number]::-webkit-outer-spin-button{-webkit-appearance:none;margin:0}input[type=number]{-moz-appearance:textfield}.wrap{position:relative;border-bottom:1px solid rgba(0,0,0,.1);border-radius:4px;height:52px;padding:0 12px;background-color:#fff}.wrap .modal-wrap{position:absolute;background-color:#fff;top:55px;left:0;border:1px solid rgba(0,0,0,.1);border-radius:4px;padding:6px}.wrap .modal-wrap .country-list{margin-top:10px;overflow:-moz-scrollbars-none;overflow-y:scroll;width:280px;max-height:250px}.wrap .modal-wrap .country-list::-webkit-scrollbar{display:none}.wrap .modal-wrap .modal-country-wrap{width:100%;padding:5px 0}.wrap .modal-wrap .modal-country-wrap:hover{background-color:#0000001a}.wrap .modal-wrap .search-box{border:none;outline:none;width:100%;border-bottom:1px solid rgba(0,0,0,.1);font-size:13px}.wrap .input-box{border:none;outline:none;width:100%;padding:0 8px}.wrap .text{font-size:16px;font-weight:400;line-height:24px;letter-spacing:-.011px}.wrap .flag{width:20px;height:auto}.active{border-bottom:1px solid #0043CE!important}.error{border-bottom:1px solid #DD3130!important}.border{border:1px solid rgba(0,0,0,.1)}.flex{display:flex}.align-center{align-items:center}.ml-5{margin-left:5px}\n"], dependencies: [{ kind: "ngmodule", type: CommonModule }, { kind: "directive", type: i1.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "pipe", type: i1.KeyValuePipe, name: "keyvalue" }, { kind: "ngmodule", type: FormsModule }, { kind: "directive", type: i2.DefaultValueAccessor, selector: "input:not([type=checkbox])[formControlName],textarea[formControlName],input:not([type=checkbox])[formControl],textarea[formControl],input:not([type=checkbox])[ngModel],textarea[ngModel],[ngDefaultControl]" }, { kind: "directive", type: i2.NumberValueAccessor, selector: "input[type=number][formControlName],input[type=number][formControl],input[type=number][ngModel]" }, { kind: "directive", type: i2.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { kind: "directive", type: i2.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.12", ngImport: i0, type: AngularPhoneNumberInput, decorators: [{
            type: Component,
            args: [{ selector: 'angular-phone-number-input', standalone: true, imports: [CommonModule, FormsModule], providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(() => AngularPhoneNumberInput),
                            multi: true,
                        },
                        {
                            provide: NG_VALIDATORS,
                            useExisting: forwardRef(() => AngularPhoneNumberInput),
                            multi: true,
                        },
                    ], template: "<div class=\"wrap flex align-center\" [class.border]=\"border\" [class.active]=\"inputFocused\" [class.error]=\"error\">\n    <div class=\"flex\">\n        @if(selectedCountry){\n        <div class=\"flex align-center\" (click)=\"toggleModal()\">\n            <span class=\"flag\" [innerHTML]=\"getIcon(selectedCountry.flag)\"></span>\n\n            <svg class=\"ml-5\" xmlns=\"http://www.w3.org/2000/svg\" width=\"10\" height=\"6\" viewBox=\"0 0 10 6\" fill=\"none\">\n                <path d=\"M4.99959 5.13698L0.725586 0.862976H9.27359L4.99959 5.13698Z\" fill=\"black\" fill-opacity=\"0.6\" />\n            </svg>\n        </div>\n        <span class=\"text ml-5\">{{selectedCountry.dialCode}}</span>\n        }@else{\n        <div class=\"flex align-center\" (click)=\"toggleModal()\">\n            <span class=\"flag\" [innerHTML]=\"getIcon(globeSVG)\"></span>\n            <svg class=\"ml-5\" xmlns=\"http://www.w3.org/2000/svg\" width=\"10\" height=\"6\" viewBox=\"0 0 10 6\" fill=\"none\">\n                <path d=\"M4.99959 5.13698L0.725586 0.862976H9.27359L4.99959 5.13698Z\" fill=\"black\" fill-opacity=\"0.6\" />\n            </svg>\n        </div>\n        }\n\n    </div>\n\n    <input class=\"input-box text\" type=\"number\" placeholder=\"\" [(ngModel)]=\"value\" [disabled]=\"disabled\"\n        (input)=\"onChange(value); onInputChanged()\" (blur)=\"onBlur(); onTouched()\" (focus)=\"onFocus()\" />\n\n    @if(showModal){\n    <div class=\"modal-wrap\">\n        <div>\n            <input class=\"search-box\" (input)=\"onSearchChange($event)\" placeholder=\"Search\" />\n        </div>\n        <div class=\"country-list\">\n            <div style=\"display: flex;\" *ngFor=\"let data of dataSet | keyvalue\">\n                <div class=\"modal-country-wrap flex\" (click)=\"selectCountry(data)\">\n                    <span class=\"flag\" [innerHTML]=\"getIcon(data.value.flag)\"></span>\n                    <span class=\"modal-country ml-5\">{{data.value.dialCode}}</span>\n                    <span class=\"modal-country ml-5\">{{data.value.country}}</span>\n                </div>\n            </div>\n        </div>\n    </div>\n    }\n</div>", styles: ["input[type=number]::-webkit-inner-spin-button,input[type=number]::-webkit-outer-spin-button{-webkit-appearance:none;margin:0}input[type=number]{-moz-appearance:textfield}.wrap{position:relative;border-bottom:1px solid rgba(0,0,0,.1);border-radius:4px;height:52px;padding:0 12px;background-color:#fff}.wrap .modal-wrap{position:absolute;background-color:#fff;top:55px;left:0;border:1px solid rgba(0,0,0,.1);border-radius:4px;padding:6px}.wrap .modal-wrap .country-list{margin-top:10px;overflow:-moz-scrollbars-none;overflow-y:scroll;width:280px;max-height:250px}.wrap .modal-wrap .country-list::-webkit-scrollbar{display:none}.wrap .modal-wrap .modal-country-wrap{width:100%;padding:5px 0}.wrap .modal-wrap .modal-country-wrap:hover{background-color:#0000001a}.wrap .modal-wrap .search-box{border:none;outline:none;width:100%;border-bottom:1px solid rgba(0,0,0,.1);font-size:13px}.wrap .input-box{border:none;outline:none;width:100%;padding:0 8px}.wrap .text{font-size:16px;font-weight:400;line-height:24px;letter-spacing:-.011px}.wrap .flag{width:20px;height:auto}.active{border-bottom:1px solid #0043CE!important}.error{border-bottom:1px solid #DD3130!important}.border{border:1px solid rgba(0,0,0,.1)}.flex{display:flex}.align-center{align-items:center}.ml-5{margin-left:5px}\n"] }]
        }], propDecorators: { error: [{
                type: Input
            }], defaultCountry: [{
                type: Input
            }], preferredCountries: [{
                type: Input
            }], border: [{
                type: Input
            }] } });
